docs/AUTH_PROTOCOL.md

```diff
@@ -0,0 +1,393 @@
+
+## Authentication Mechanisms and Conversation
+
+Authentication mechanisms are standardized UI patterns that an authenticator can request from a frontend. For instance: if the authenticator asks the frontend for a password, the frontend will pick an appropriate UI for password entry. Mechanisms can be categorized as "primary" or "secondary". There can only be one primary mechanism active (i.e. visible on screen) at a time, so if the authenticator asks to show a different primary mechanism the previous UI should be replaced with the new one. Secondary mechanisms can co-exist with the primary mechanism and with each-other
+
+Once a flow is started, the authentication "conversation" begins. At a high level, a conversation looks like the following:
+
+1. The authenticator requests a primary authentication mechanism from the frontend
+2. The frontend displays the appropriate UI for this authentication mechanism
+3. The user interacts with the UI to give it some data (types in a password, touches a fingerprint sensor, etc)
+4. Once the user is finished, the frontend sends the user's entered data to the authenticator
+5. If the user has failed to authenticate, or if there are no more steps left in the flow, the authenticator ends the conversation and returns success or failure
+6. Otherwise, the authenticator continues with the next step of the flow. Return to step 1
+
+Note that the use of this protocol encapuslates one authentication attempt. A flow might have multiple steps in it (i.e. "password" followed by "otp"), and the authenticator's ability to switch primary authentication mechansims is designed just for that. The authenticator should not handle a primary authentication method's failure on its own. For example: if the user enters an incorrect password, the authenticator should not prompt for a second attempt. Instead, it should emit the `authenticationFailed` event and terminate the conversation. Conversely, the authenticator _is_ expected to handle secondary authentication method failures on its own, because these are not fatal to the overall login attempt.
+
+Each authentication mechanism is started via an event that the authenticator sends. The following is a description of each such event and what the frontend should do in response:
+
+- `password`: A generic password box
+    - This method is always a primary method
+    - Arguments (data that the authenticator includes with the event it sends to the frontend):
+        - `overridePrompt`: Overrides the default prompt in the password entry box. Optional. Should only be used if absolutely necessary
+    - Response: frontend emits the `response` event with the following arguments:
+        - `password`: The password entered by the user
+
+- `text`: Arbitrary text input
+    - This method is always primary
+    - Arguments:
+        - `prompt`: The prompt to put in the text box. Required
+    - Response: `response` event with arguments:
+        - `text`: The text entered by the user
```

Comment:
Any reason why they are not merged as they used to be (as `entry` visible or not)?

Also in authd we even support `digit` only, so while I wouldn't go as far as providing a regex to validate the input, maybe we can just add a flag for only digits (for numeric PINs).

EDIT: Saw that PIN was handled later, but not sure if we want to merge... Although I see the point of having them separated for future updates.